packages/x-date-pickers/src/internals/utils/date-utils.ts

```diff
@@ -142,3 +143,45 @@ export const getTodayDate = <TDate>(utils: MuiPickersAdapter<TDate>, valueType:
 const dateViews = ['year', 'month', 'day'];
 export const isDatePickerView = (view: DateOrTimeViewWithMeridiem): view is DateView =>
   dateViews.includes(view);
+
+export const getDatePickerFormatFromViews = (
+  utils: MuiPickersAdapter<any>,
+  { format, views }: { format?: string; views: readonly DateView[] },
+  isInToolbar: boolean,
+) => {
+  if (format != null) {
+    return format;
+  }
+
+  const formats = utils.formats;
+  if (areViewEquals(views, ['year'])) {
+    return formats.year;
+  }
+
+  if (areViewEquals(views, ['month'])) {
+    return formats.month;
+  }
+
+  if (areViewEquals(views, ['day'])) {
+    return formats.dayOfMonth;
+  }
+
+  if (areViewEquals(views, ['month', 'year'])) {
+    return `${formats.month} ${formats.year}`;
+  }
+
+  if (areViewEquals(views, ['day', 'month'])) {
+    return `${formats.month} ${formats.dayOfMonth}`;
+  }
+
+  if (isInToolbar) {
+    // Little localization hack (Google is doing the same for android native pickers):
+    // For english localization it is convenient to include weekday into the date "Mon, Jun 1".
+    // For other locales using strings like "June 1", without weekday.
+    return /en/.test(utils.getCurrentLocaleCode())
+      ? formats.normalDateWithWeekday
+      : formats.normalDate;
+  }
+
+  return formats.keyboardDate;
```

Comment:
I'm not hardcoding a format based on `day` / `month` and `year` because the locale can change the order and formatting of those elements (which is very unlikely for the time and was not handled by our default formats which do not use any meta token).